src/sagemaker_sklearn_extension/feature_extraction/sequences.py

```diff
@@ -0,0 +1,414 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#      http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import numpy as np
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_array, check_is_fitted
+from tsfresh import extract_features
+from tsfresh.feature_extraction import ComprehensiveFCParameters
+from tsfresh.feature_extraction import EfficientFCParameters
+from tsfresh.feature_extraction import MinimalFCParameters
+from tsfresh.utilities.dataframe_functions import impute
+
+from sagemaker_sklearn_extension.preprocessing.data import RobustStandardScaler
+
+
+class TimeSeriesFeatureExtractor(BaseEstimator, TransformerMixin):
+    """Wrap TSFlattener and TSFreshFeatureExtractor to extract time series features from multiple sequence columns.
+
+    The input is an array where rows are observations and columns are sequence features. Each sequence feature is a
+    string containing a sequence of comma-separate numbers.
+
+    For each column, TSFlattener extracts numerical values from the strings and returns a list of np.arrays as output,
+    and then TSFreshFeatureExtractor extracts time series features from each list. The outputs from each column are then
+    stacked horizontally into a single array.
+
+    See TSFlattener and TSFreshFeatureExtractor for more details.
+
+    Parameters
+    ----------
+    max_allowed_length : int (default = 10000)
+        Maximum allowed length of an input sequence. If the length of a sequence is greater than ``max_allowed_length``,
+        the transformer will truncate its beginning or end as indicated by ``trim_beginning``.
+
+    trim_beginning : bool (default = True)
+        If a sequence length exceeds ``max_allowed_length``, trim its start and only keep the last `max_allowed_length``
+        values if ``trim_beginning`` = True, otherwise trim the end and only keep the first max_allowed_length values.
+
+    augment : boolean (default=True):
+        Whether to append the tsfresh features to the original data (if True),
+        or output only the extracted tsfresh features (if False).
+        If True, also pad shorter sequences (if any) in the original data with np.nans, so that all sequences
```

Comment:
Done. Changed the default to False as it also works better in our experiments while using fewer features. Even with augment = True though NaNs are not an issue, since we interpolate right after padding ensuring that no NaNs are fed to the downstream algorithms.